src/gradient_descent.c

```diff
@@ -0,0 +1,380 @@
+
+    // Form a more precise bracket for the minimum
+    if (fCurrent <= fXMin) {
+      if (current >= *xMin) {
+        a = *xMin;
+      } else {
+        b = *xMin;
+      }
+
+      prevSecondXMin = secondXMin;
+      secondXMin = *xMin;
+      *xMin = current;
+
+      fPrevSecondXMin = fSecondXMin;
+      fSecondXMin = fXMin;
+      fXMin = fCurrent;
+    } else {
+      if (current < *xMin) {
+        a = current;
+      } else {
+        b = current;
+      }
+      if (fCurrent <= fSecondXMin || secondXMin == *xMin) {
+        prevSecondXMin = secondXMin;
+        secondXMin = current;
+        fPrevSecondXMin = fSecondXMin;
+        fSecondXMin = fCurrent;
+      } else if (fCurrent <= fPrevSecondXMin || prevSecondXMin == *xMin ||
+                 prevSecondXMin == secondXMin) {
+        prevSecondXMin = current;
+        fPrevSecondXMin = fCurrent;
+      }
+    }
+  }
+
+  // We should not reach this point to many iterations
+  return fXMin;
+}
+
+/// @brief Initially brackets the minimum between 3 points (a, b, c). Uses
+/// parabolic interpolation to find the bracket
+/// @param func Function transforming an N dimension function to a 1 dimension
+/// function
+/// @param f Function we want to transfor to a 1 dimension function
+/// @param initialPoint Initial point to base the movement of
+/// @param direction Direction in which to move
+/// @param n Number of dimensions of the original function
+/// @return Structure containing the bracket for the minimum
+static Bracket bracketMinimum(f1dimension func, function f,
+                              gradient_real initialPoint[],
+                              gradient_real direction[], int n) {
+  // Set a, b and c at arbitrary initial values
+  Bracket bracket = {0.0, 1.0, 0.0};
+
+  // Evaluate function at initial points a and b
+  gradient_real fa = (*func)(f, bracket.a, initialPoint, direction, n);
+  gradient_real fb = (*func)(f, bracket.b, initialPoint, direction, n);
+
+  // We want to go downhill from a to b to bracket the minimum
+  // so if fb > fa we need to swap them
+  if (fb > fa) {
+    gradient_real temp = bracket.a;
+    bracket.a = bracket.b;
+    bracket.b = temp;
+
+    temp = fa;
+    fa = fb;
+    fb = fa;
+  }
+
+  // Initial guess for c
+  bracket.c = (bracket.b) + GOLD * (bracket.b - bracket.a);
+  gradient_real fc = (*func)(f, bracket.c, initialPoint, direction, n);
+
+  // Loop until we bracket the minimum
+  while (fb > fc) {
+    // Compute u the inflexion point of the parabola defined by
+    // points a, b and c
+    gradient_real r = (bracket.b - bracket.a) * (fb - fc);
+    gradient_real q = (bracket.b - bracket.c) * (fb - fa);
+    gradient_real diff = q - r;
+    if (diff == 0) {
+      diff = EPS; // Prevent division by 0
+    }
+    gradient_real inflexion = (bracket.b) - ((bracket.b - bracket.c) * q -
+                                             (bracket.b - bracket.a) * r) /
+                                                (2.0 * diff);
```

Comment:
Please correct me if I am wrong but it seems like diff is not updated insided the while loop. If that is the case it would be better to define the inverse of 2.0 * diff outside this loop. That way you can use the inverse and reduce the amount of divisions

Reply:
Line 204

Reply:
As mentionned by @felix642 Diff is defined inside the while loop.